prbt_hardware_support/test/unittests/unittest_speed_observer.cpp

```diff
@@ -0,0 +1,406 @@
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <thread>
+
+#include <gmock/gmock.h>
+
+#include <ros/ros.h>
+#include <std_srvs/SetBool.h>
+#include <tf2_ros/transform_broadcaster.h>
+
+#include <pilz_testutils/async_test.h>
+#include <prbt_hardware_support/FrameSpeeds.h>
+#include <prbt_hardware_support/SetSpeedLimit.h>
+#include <prbt_hardware_support/ros_test_helper.h>
+#include <prbt_hardware_support/speed_observer.h>
+#include <prbt_hardware_support/wait_for_service.h>
+
+namespace speed_observer_test
+{
+using ::testing::_;
+using ::testing::DoAll;
+using ::testing::Return;
+using ::testing::SetArgReferee;
+
+using namespace prbt_hardware_support;
+
+static const std::string BARRIER_SLOW{ "BARRIER_SLOW" };
+static const std::string BARRIER_FAST{ "BARRIER_FAST" };
+static const std::string BARRIER_LIMIT{ "BARRIER_LIMIT" };
+static const std::string BARRIER_LIMIT_LOW{ "BARRIER_LIMIT_LOW" };
+
+static const std::string FRAME_SPEEDS_TOPIC_NAME{ "/frame_speeds" };
+static const std::string STOP_TOPIC_NAME{ "/safe_torque_off" };
+static const std::string TEST_BASE_FRAME{ "test_base" };
+static const std::string TEST_FRAME_A{ "a" };
+static const std::string TEST_FRAME_B{ "b" };
+static const double TEST_FREQUENCY{ 20 };
+static const double SQRT_2_HALF{ 1 / sqrt(2) };
+
+class SpeedObserverUnitTest : public testing::Test, public testing::AsyncTest
+{
+public:
+  void SetUp() override;
+  void TearDown() override;
+
+  MOCK_METHOD1(frame_speeds_cb_mock, void(const FrameSpeeds::ConstPtr& msg));
+  MOCK_METHOD2(stop_cb_mock, bool(std_srvs::SetBool::Request& req, std_srvs::SetBool::Response& res));
+  void publishTfAtSpeed(double v);
+  void stopTfPublisher();
+
+protected:
+  ros::Subscriber speed_subscriber_;
+  ros::ServiceServer stop_subscriber_;
+  ros::NodeHandle nh_;
+  ros::NodeHandle pnh_{ "~" };
+  std::vector<std::string> additional_frames_;
+  bool tf_publisher_running_{ false };
+};
+
+void SpeedObserverUnitTest::SetUp()
+{
+  speed_subscriber_ =
+      nh_.subscribe<FrameSpeeds>(FRAME_SPEEDS_TOPIC_NAME, 1, &SpeedObserverUnitTest::frame_speeds_cb_mock, this);
+  stop_subscriber_ = nh_.advertiseService(STOP_TOPIC_NAME, &SpeedObserverUnitTest::stop_cb_mock, this);
+
+  waitForService(STOP_TOPIC_NAME);
+}
+
+void SpeedObserverUnitTest::TearDown()
+{
+}
+
+void SpeedObserverUnitTest::publishTfAtSpeed(double v)
+{
+  static tf2_ros::TransformBroadcaster br;
+  ros::Rate r = ros::Rate(TEST_FREQUENCY * 3);  // publishing definitely faster then observing
```

Comment:
If you increase the frequence does the noise reduce? I see you have to cover quite a range in your expectations

Reply:
not really